Begin testing models from the ONNX Model Zoo.

ScottTodd opened the pull request:
Progress on https://github.com/iree-org/iree-test-suites/issues/6.

These new tests

* Downloads models from https://github.com/onnx/models
* Extracts metadata from the models to determine which functions to call with random data
* Runs the models through [ONNX Runtime](https://onnxruntime.ai/) as a reference implementation
* Imports the models using `iree-import-onnx` (until we have a better API: https://github.com/iree-org/iree/issues/18289)
* Compiles the models using `iree-compile` (currently just for `llvm-cpu` but this could be parameterized later)
* Runs the models using `iree-run-module`, checking outputs using `--expected_output` and the reference data

Tests are written in Python using a set of pytest helper functions. As the tests run, they can log details about what commands they are running. When run locally, a `artifacts/` directory will contain all the relevant files. More can be done in follow-up PRs to improve the ergonomics there (like generating flagfiles and grouping files by test name/type).

Each test case can use XFAIL like `@pytest.mark.xfail(raises=IreeRunException)`. As we test across multiple backends or want to configure the test suite from another repo (e.g. [iree-org/iree](https://github.com/iree-org/iree)), we can explore more expressive marks.

Note that unlike the ONNX _operator_ tests, these tests use `onnxruntime` and `iree-import-onnx` at test time. The operator tests handle that as an infrequently ran offline step. We could do something similar here, but the test inputs and outputs can be rather large for real models and that gets into Git LFS or cloud storage territory.

If this test authoring model works well enough, we can do something similar for other ML frameworks like TFLite (https://github.com/iree-org/iree-test-suites/issues/5).

ScottTodd reviewed (commented):
Some self review comments from weekend thinking. Busy for a bit then will make those changes

ScottTodd commented on onnx_models/conftest.py:
```diff
@@ -0,0 +1,355 @@
+        logger.debug("ONNX model metadata:")
+        logger.debug(onnx_model_metadata)
+
+        # Run through ONNX Runtime.
+        onnx_session = InferenceSession(upgraded_onnx_path)
+        output_names = [output["name"] for output in onnx_model_metadata["outputs"]]
+        inputs = {}
+        for input in onnx_model_metadata["inputs"]:
+            inputs[input["name"]] = input["input_data"]
+        onnx_results = onnx_session.run(output_names, inputs)
```
Would also be nice to get the files in some format that is amenable to manual/automated benchmarking. Hmmm

ScottTodd replied:
Added some dataclasses. Will try to not overgeneralize now, but could share those with TF / TFLite / PyTorch / JAX tests too...

zjgarvey commented:
> Ping @zjgarvey ? Any high level feedback? I should have some time this week to address my own comments, but I'd like to prefetch any other feedback/discussion.

I'll take a look today!

zjgarvey reviewed (commented):
My biggest concern right now is the handling of dynamic dims and getting the input/output signatures to work in those cases. 

Some suggestions pertaining to this:

1. Have a test-specific dictionary of `dim_param` assignments to handle models with dynamic dims.
2. Since you are already generating an onnxruntime session, it might be easier to pass the session itself, together with a dim_param dictionary, to generate the input signature.
3. Again, since you are using the onnxruntime session, the most reliable method of getting a correct output signature is definitely from the reference outputs of that session run. It might be prohibitively complicated to try and infer the output shapes from a dynamic onnx model. Even with shape inference and dim_params being included, I'm not sure it would be able to determine correct output shapes without actually fixing the dim_params in the onnx model itself (however, there is tooling for that <https://github.com/microsoft/onnxruntime/blob/291a5352b27ded5714e5748b381f2efb88f28fb9/tools/python/util/onnx_model_utils.py#L177>).


If I don't revisit this soon, please feel free to ping again.

ScottTodd commented on onnx_models/conftest.py:
```diff
@@ -0,0 +1,355 @@
+    #   A) List all metadata explicitly
+    #   B) Get metadata on demand from the .onnx protobuf using 'onnx'
+    #   C) Get metadata on demand from the InferenceSession using 'onnxruntime'
+    # This is option B.
```
Another option I'm considering is having test case import be an offline, tool-assisted step, instead of an online step that always runs as part of the tests. That way, we can have developers importing model tests decide explicitly what functions and signatures they want to test. It would be nice if there was only one obvious function and signature for each model that can be safely inferred though :)

ScottTodd replied:
Made some functional updates, but I'll continue to iterate on the code style.

This is now using option C - getting the input and output signatures from `onnxruntime.InferenceSession`. The output signatures are now exactly what comes from onnxruntime, while the input signatures are still turning dynamic dims into 1.

ScottTodd commented on onnx_models/conftest.py:
```diff
@@ -0,0 +1,355 @@
+        f"Importing '{onnx_path.relative_to(THIS_DIR)}' to '{imported_mlir_path.relative_to(THIS_DIR)}'"
+    )
+    exec_args = [
+        "iree-import-onnx",
```
Not yet. That's part of https://github.com/iree-org/iree/issues/18289

ScottTodd commented on onnx_models/conftest.py:
```diff
@@ -0,0 +1,355 @@
+        for input in onnx_model_metadata["inputs"]:
+            input_type = input["iree_type"]
+            input_data_path = input["input_data_path"]
+            run_module_args.append(f"--input={input_type}=@{input_data_path}")
```
We have control over the inputs. As long as there aren't restrictions (like must be a multiple of 4, or two dynamic inputs must be equal), setting to 1 seems safe enough to start. We could also override per test case or add a flag that lets you override for the entire test suite.

I at least made it so the output shape comes from the inference session, since that value is out of out control.

zjgarvey replied:
Yeah, I know many llms have dim params like `past_seq_len` and then elsewhere `past_seq_len + 1`, so there are definitely restrictions for some models. 

After looking more carefully at the setup (which is really nice, btw), I think it would be pretty feasible to just add an optional `dim_param_dict` as an input for `compare_between_iree_and_onnxruntime_fn`.

zjgarvey replied:
In short, I think that could be added later as necessary.

ScottTodd replied:
Yep. I just started with vision classification models but as more are added this test code will grow more features. Are there particular ONNX models that you know need that treatment? I can look at importing those next.

ScottTodd commented on onnx_models/utils.py:
```diff
@@ -0,0 +1,185 @@
+    )
+
+
+def convert_node_arg_type_to_numpy_dtype(type: str):
+    # TODO(scotttodd): use onnx.TensorProto instead? enums > strings
+    if type == "tensor(float)":
+        return np.float32
+    raise NotImplementedError(f"type conversion for '{type}' not implemented")
```
Seems like the type here is just a string, but I can't find a complete list of what types are available: https://onnxruntime.ai/docs/api/python/api_summary.html#onnxruntime.NodeArg 

I'm tempted to use the TensorProto enums, but that would then involve loading the model into memory again.

ScottTodd commented:
Okay, I made a bunch of updates today and I'm happy enough with where this is now to merge it and start running it nightly.

Major changes:

* Moved more util code from conftest.py to utils.py
* Reworked metadata extraction and reference output generation to fully use `onnxruntime`, not `onnx` (Protobuf)
* Imported more vision classification tests (still only f32)
* Organized tests into a subfolder
* Improved documentation and gave report generation a try (https://scotttodd.github.io/iree-test-suites/onnx_models/report_2024_09_17.html)

PTAL?

zjgarvey commented on onnx_models/conftest.py:
```diff
@@ -0,0 +1,243 @@
+        logger.debug(f"  iree parameter: {iree_type}")
+
+        # Create a numpy tensor with some random data for the input.
+        numpy_dtype = convert_node_arg_type_to_numpy_dtype(input.type)
+        if numpy_dtype == np.float32 or numpy_dtype == np.float64:
+            input_data = rng.random(numpy_dimensions, dtype=numpy_dtype)
+        elif numpy_dtype == np.int32 or numpy_dtype == np.int64:
+            input_data = rng.integers(numpy_dimensions, dtype=numpy_dtype)
+        else:
+            raise NotImplementedError(f"Unsupported numpy type: {numpy_dtype}")
```
I'm sure you've already looked at this, but could consider borrowing some of <https://github.com/nod-ai/SHARK-TestSuite/blob/dbbe625d5b7beac0b4465f7b90441d9f4593bc57/alt_e2eshark/e2e_testing/onnx_utils.py#L31>

zjgarvey replied:
Especially since you aren't using the `numpy_dtype` elsewhere.

ScottTodd replied:
Ah! I hadn't looked there recently. Thanks for the reminder. Will pull in some of the cases from there.

ScottTodd replied:
Refactored a bit, pulling more code from there. Thanks again!

zjgarvey commented:
@ScottTodd , I don't have any pressing comments. The printout you shared was really easy to navigate and the code here feels like it will be painless to iterate on with updates. Please let me know when you feel like this is in a state to merge, and I'll take a final look.
